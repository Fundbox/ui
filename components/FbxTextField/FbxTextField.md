# FbxTextField

## Basic usage

- Cool description

```html
<fbx-text-field
  name="email"
  class="input"
  validations="required|email"
  placeholder="Enter your email address"
  :label="'Email Address'"
  v-model.trim="email"
  data-qa="login-email"
/>
```
