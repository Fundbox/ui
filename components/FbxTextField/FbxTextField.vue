<template>
  <div class="fbx-text-field">
    <label class="fbx-text-field__label">{{ label }}</label>

    <input
      :type="type"
      tabindex="0"
      class="fbx-text-field__input"
      :class="{ invalid: isInvalid }"
      v-validate="validations"
      v-bind="$attrs"
      :value="value"
      @input="onInput"
      @change="onChange"
    />

    <fbx-validation-message class="validation-message" v-if="isInvalid">{{ validationMessage }}</fbx-validation-message>
  </div>
</template>

<script>
export default {
  name: "FbxTextField",
  inheritAttrs: false,
  inject: ["$validator"],
  data() {
    return {
      type: this.$attrs.type || "text",
    }
  },
  props: {
    label: String,
    value: String,
    validations: String,
  },
  computed: {
    isInvalid() {
      return this.errors.has(this.$attrs.name)
    },
    validationMessage() {
      return this.errors.first(this.$attrs.name)
    }
  },
  methods: {
    onInput(event) {
      this.$emit("input", event.target.value)
      // this.$forceUpdate()
    },
    onChange(event) {
      this.$emit("change", event.target.value)
    },
  },
}
</script>

<style lang="scss" scoped>
@import "./../styles/utils/color-palette";
@import "./../styles/utils/mixins";

.fbx-text-field {
  .fbx-text-field__label {
    @include font(10);
  }

  .fbx-text-field__input {
    width: inherit;
    height: 50px;
    min-width: 240px;
    padding: 15px;
    @include font(16);
    border-radius: 3px;
    border: none;
    background-color: $extra-light-gray;
    outline: none;

    &:focus {
      border-bottom: 2px solid $dark-green;
    }

    &.invalid {
      background-color: $extra-light-red;
      border-bottom: 2px solid $light-red;
    }
  }

  .validation-message {
    margin-top: 10px;
  }
}
</style>
