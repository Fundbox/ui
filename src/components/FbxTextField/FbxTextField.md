# FbxTextField

## Basic usage

- Use `FbxTextField` for inputs of common types, e.g. `type="text"`, `type="email"`. If your input type isn't supported, open up a PR to add support for that input type
- The default `type` is `text`, if you don't provide one
- You can pass a `mask` prop to set up masking. See more info [here](https://www.npmjs.com/package/v-mask)

ðŸš¨Note: you cannot currently use both `clearable` and `type="password"` as the Show/Hide and X buttons overlap ðŸš¨

```html
<!-- Example Usage -->
<fbx-text-field
  name="email"
  class="email-field"
  validations="required|email"
  placeholder="Enter your email address"
  :label="'Email Address'"
  v-model.trim="email"
  data-qa="my-email-qa"
/>

<!-- Example Usage with mask -->
<fbx-text-field
  mask="(###) ###-####"
  name="phone"
  class="input"
  validations="required"
  placeholder="Enter your phone number"
  :label="labelText"
  v-model="inputText"
/>

<!-- Example Usage with address -->
<fbx-text-field
  address-autocomplete
  :validations="{
    addressRequired: addressData,
    addressInsideUSA: addressData,
    fullAddress: addressData
  }"
  name="phone"
  class="input"
  placeholder="Enter your address"
  v-model="inputText"
  @addressDataChanged="onAddressDataChanged"
/>

<!-- Example Usage with a clear icon -->
<fbx-text-field
  validations="required"
  name="phone"
  class="input"
  placeholder="Enter your search"
  v-model="inputText"
  clearable
/>
```
