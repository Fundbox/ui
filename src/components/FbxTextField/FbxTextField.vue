<template>
  <div class="fbx-text-field">
    <label class="fbx-text-field__label">{{ label }}</label>
    <div class="fbx-text-field__wrapper">
      <div class="fbx-text-field__input-wrapper">
        <input
          ref="fbxTextFieldInput"
          v-fbx-address-autocomplete="addressAutocomplete"
          v-fbx-autofocus="autofocus"
          v-mask="mask"
          :type="type"
          tabindex="0"
          class="fbx-text-field__input"
          :class="{ password: isPassword, invalid: isInvalid, clearable: clearable }"
          v-validate="validations"
          v-bind="$attrs"
          :value="value"
          @input="onInput"
          @change="onChange"
        />

        <span class="fbx-text-field__password-button" @click="togglePassword" v-if="isPassword">{{ passwordButtonText }}</span>

        <span class="fbx-text-field__clear-icon" @click="clearField" v-if="clearable"></span>
      </div>
      <fbx-validation-message class="validation-message" v-if="isInvalid">{{ validationMessage }}</fbx-validation-message>
    </div>
  </div>
</template>

<script>
import '../../validations'
import { VueMaskDirective } from 'v-mask'
import FbxValidationMessage from '../FbxValidationMessage/FbxValidationMessage.vue'
import FbxAddressAutocomplete from '../../directives/FbxAddressAutocomplete/FbxAddressAutocomplete'
import FbxAutofocus from '../../directives/FbxAutofocus/FbxAutofocus'

export default {
  name: 'FbxTextField',
  components: {
    FbxValidationMessage
  },
  directives: {
    mask: VueMaskDirective,
    FbxAddressAutocomplete,
    FbxAutofocus
  },
  inheritAttrs: false,
  inject: ['$validator'],
  data() {
    return {
      isPassword: this.$attrs.type === 'password',
      type: this.$attrs.type || 'text'
    }
  },
  props: {
    label: String,
    value: {
      type: [String, Number],
    },
    validations: {
      type: [String, Object],
    },
    mask: {
      type: String,
      default: ''
    },
    autofocus: {
      type: Boolean,
      default: false
    },
    addressAutocomplete: {
      type: Boolean,
      default: false
    },
    clearable: {
      type: Boolean,
      default: false
    },
  },
  computed: {
    passwordButtonText() {
      return this.isPassword && this.type === 'text' ? 'Hide' : 'Show'
    },
    isInvalid() {
      return this.errors.has(this.$attrs.name, this.$attrs.scope)
    },
    validationMessage() {
      return this.errors.first(this.$attrs.name, this.$attrs.scope)
    }
  },
  methods: {
    togglePassword() {
      this.type = this.type === 'password' ? 'text' : 'password'
    },
    clearField() {
      this.$refs.fbxTextFieldInput.focus()
      this.$emit('input', '')
    },
    onInput(event) {
      this.$emit('input', event.target.value)
    },
    onChange(event) {
      this.$emit('change', event.target.value)
    },
  },
}
</script>

<style lang="scss" scoped>
@import "../../styles/utils/color-palette";
@import "../../styles/utils/mixins";

.fbx-text-field {
  position: relative;
  display: inline-block;
  width: 100%;

  .fbx-text-field__label {
    display: block;
    margin-bottom: 7px;
    color: $extra-dark-gray;
    @include font(16, light);
  }

  .fbx-text-field__wrapper {
    margin-bottom: 22px;
  }

  .fbx-text-field__input-wrapper {
    position: relative;
  }

  .fbx-text-field__input {
    display: block;
    width: 100%;
    height: 50px;
    min-width: 240px;
    padding: 15px;
    @include font(16);
    border: none;
    color: $medium-blue;
    background-color: $extra-light-gray;
    outline: none;

    &.password {
      padding-right: 60px;
    }

    &:focus {
      border-bottom: 1px solid $dark-green;
    }

    &.invalid {
      background-color: $extra-light-red;
      border-bottom: 1px solid $light-red;
    }

    &.clearable {
      padding-right: 35px;
    }
  }

  .fbx-text-field__password-button,
  .fbx-text-field__clear-icon {
    position: absolute;
    right: 15px;
    top: 50%;
    transform: translateY(-50%);
    cursor: pointer;
    user-select: none;
  }

  .fbx-text-field__password-button {
    @include font(16);
    color: $dark-green;
  }

  .fbx-text-field__clear-icon {
    right: 5px;
    display: block;
    width: 30px;
    height: 100%;
    color: $medium-blue;
    background-repeat: no-repeat;
    background-position: 50% 50%;
    background-size: 12px 10px;
    background-image: url("../../assets/icon_active_x_icon.png");

    &:hover {
      background-image: url("../../assets/icon_active_x_icon_active.png");
    }
  }

  .validation-message {
    margin-top: 10px;
  }
}
</style>
