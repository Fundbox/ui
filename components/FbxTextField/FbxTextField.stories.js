import { storiesOf } from '@storybook/vue';
import { action } from '@storybook/addon-actions';
import { withInfo } from 'storybook-addon-vue-info';
import { text } from '@storybook/addon-knobs';

import FbxTextField from '../FbxTextField';
import summary from './FbxTextField.md';
const stories = storiesOf('Components/TextField', module);
const withSummery = withInfo({ summary });

const defaultStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: "",
      labelText: text("Label", "Email Address"),
    }
  },
  watch: {
    inputText(value) {
      action(`New value: ${value}`)()
    },
  },
  template: `
    <fbx-text-field
      name="email"
      class="input"
      validations="required"
      placeholder="Enter your email"
      :label="labelText"
      v-model="inputText"
    />
  `,
});

const passwordStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: "",
      labelText: text("Label", "Password"),
    }
  },
  watch: {
    inputText(value) {
      action(`New value: ${value}`)()
    },
  },
  template: `
    <div style="width: 300px; padding: 30px;">
      <fbx-text-field
        name="password"
        type="password"
        class="input"
        validations="required"
        placeholder="Enter your password"
        :label="labelText"
        v-model="inputText"
      />
    </div>
  `,
});

const maskStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: "",
      labelText: text("Label", "Email Address"),
    }
  },
  watch: {
    inputText(value) {
      action(`New value: ${value}`)()
    },
  },
  template: `
    <fbx-text-field
      mask="(###) ###-####"
      name="phone"
      class="input"
      validations="required"
      placeholder="Enter your phone number"
      :label="labelText"
      v-model="inputText"
    />
  `
});

const autofocusStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: "",
      labelText: text("Label", "Email Address"),
    }
  },
  watch: {
    inputText(value) {
      action(`New value: ${value}`)()
    },
  },
  template: `
    <fbx-text-field
      autofocus
      name="phone"
      class="input"
      validations="required"
      placeholder="Enter your phone number"
      :label="labelText"
      v-model="inputText"
    />
  `
});

const addressAutocomplete = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: "",
      addressData: {}
    }
  },
  methods: {
    onAddressDataChanged(addressData) {
      this.addressData = addressData;
    }
  },
  watch: {
    inputText(value) {
      action(`New value`)(value)
    },
    addressData(value) {
      action(`Address data`)(value)
    }
  },
  template: `
    <fbx-text-field
      address-autocomplete
      :validations="{
        addressRequired: addressData,
        addressInsideUSA: addressData,
        fullAddress: addressData
      }"
      name="phone"
      class="input"
      placeholder="Enter your address"
      v-model="inputText"
      @addressDataChanged="onAddressDataChanged"
    />
  `
});

stories.add('default', withSummery(defaultStory));
stories.add('password', withSummery(passwordStory));
stories.add('mask', withSummery(maskStory));
stories.add('autofocus', withSummery(autofocusStory));
stories.add('address autocomplete', withSummery(addressAutocomplete));
