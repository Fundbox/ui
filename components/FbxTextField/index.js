import Vue from "vue";
import FbxTextField from "./FbxTextField.vue";
Vue.component("FbxTextField", FbxTextField);
export default FbxTextField;
