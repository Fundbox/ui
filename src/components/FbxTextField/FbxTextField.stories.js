import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import { withInfo } from 'storybook-addon-vue-info'
import { text } from '@storybook/addon-knobs'

import FbxTextField from './FbxTextField.vue'
import summary from './FbxTextField.md'
const stories = storiesOf('Components/TextField', module)
const withSummary = withInfo({ summary })

const defaultStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: '',
      labelText: text('Label', 'Email Address'),
    }
  },
  watch: {
    inputText(value) {
      action(`New value: ${value}`)()
    },
  },
  template: `
    <fbx-text-field
      name="email"
      class="input"
      validations="required|email"
      placeholder="Enter your email"
      :label="labelText"
      :data-qa="email-field-input"
      v-model="inputText"
    />
  `,
})

const passwordStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: '',
      labelText: text('Label', 'Password'),
    }
  },
  watch: {
    inputText(value) {
      action(`New value: ${value}`)()
    },
  },
  template: `
    <div style="width: 300px; padding: 30px;">
      <fbx-text-field
        name="password"
        type="password"
        class="input"
        validations="required"
        placeholder="Enter your password"
        :label="labelText"
        v-model="inputText"
      />
    </div>
  `,
})

const maskStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: '',
      labelText: text('Label', 'Email Address'),
    }
  },
  watch: {
    inputText(value) {
      action(`New value: ${value}`)()
    },
  },
  template: `
    <fbx-text-field
      mask="(###) ###-####"
      name="phone"
      class="input"
      validations="required"
      placeholder="Enter your phone number"
      :label="labelText"
      v-model="inputText"
    />
  `
})

const autofocusStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: '',
      labelText: text('Label', 'Email Address'),
    }
  },
  watch: {
    inputText(value) {
      action(`New value: ${value}`)()
    },
  },
  template: `
    <fbx-text-field
      autofocus
      name="phone"
      class="input"
      validations="required"
      placeholder="Enter your phone number"
      :label="labelText"
      v-model="inputText"
    />
  `
})

const addressAutocompleteStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: '',
      addressData: {}
    }
  },
  methods: {
    onAddressDataChanged(addressData) {
      this.addressData = addressData
    }
  },
  watch: {
    inputText(value) {
      action(`New value`)(value)
    },
    addressData(value) {
      action(`Address data`)(value)
    }
  },
  template: `
    <fbx-text-field
      address-autocomplete
      :validations="{
        addressRequired: addressData,
        addressInsideUSA: addressData,
        fullAddress: addressData
      }"
      name="phone"
      class="input"
      placeholder="Enter your address"
      v-model="inputText"
      @addressDataChanged="onAddressDataChanged"
    />
  `
})

const clearableStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: '',
    }
  },
  watch: {
    inputText(value) {
      action(`New value`)(value)
    },
  },
  template: `
    <div style="width: 300px;">
      <fbx-text-field
        name="search"
        class="input"
        placeholder="Enter your search"
        v-model="inputText"
        clearable
      />
    </div>
  `,
})

const currencyStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: '',
      inputText2: '9999.99',
      inputText3: '',
    }
  },
  watch: {
    inputText(value) {
      action(`New value`)(value)
    },
    inputText2(value) {
      action(`New value`)(value)
    },
    inputText3(value) {
      action(`New value`)(value)
    },
  },
  template: `
    <div style="width: 400px;">
      <fbx-text-field
        name="amount"
        class="input"
        placeholder="Enter an amount"
        v-model="inputText"
        currency
      />
      <fbx-text-field
        label="Prefilled"
        name="amount2"
        class="input"
        placeholder="Enter an amount"
        v-model="inputText2"
        currency
      />
      <fbx-text-field
        label="With Validations (min $1,000, max $1,999.99)"
        name="amount3"
        class="input"
        placeholder="Enter an amount"
        v-model="inputText3"
        currency
        :validations="{
          required: 'true',
          max_value: 1999.99,
          min_value: 1000,
        }"
      />
    </div>
  `,
})

const editableStory = () => ({
  components: { FbxTextField },
  data() {
    return {
      inputText: '',
      inputText2: '',
    }
  },
  watch: {
    inputText(value) {
      action(`New value`)(value)
    },
    inputText2(value) {
      action(`New value`)(value)
    },
  },
  template: `
    <div style="width: 400px;">
      <fbx-text-field
        name="word"
        class="input"
        placeholder="Enter a word"
        v-model="inputText"
        editable
      />
      <fbx-text-field
        label="With currency and validations (min $1,000, max $9,999.99)"
        name="amount"
        class="input"
        placeholder="Enter an amount"
        v-model="inputText2"
        editable
        currency
        :validations="{
          required: 'true',
          max_value: 9999.99,
          min_value: 1000,
        }"
      />
    </div>
  `,
})

stories.add('default', withSummary(defaultStory))
stories.add('password', withSummary(passwordStory))
stories.add('mask', withSummary(maskStory))
stories.add('autofocus', withSummary(autofocusStory))
stories.add('address autocomplete', withSummary(addressAutocompleteStory))
stories.add('clearable', withSummary(clearableStory))
stories.add('currency', withSummary(currencyStory))
stories.add('editable', withSummary(editableStory))
