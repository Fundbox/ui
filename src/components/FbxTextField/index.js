import FbxTextField from './FbxTextField.vue'
import { registerComponents, vueUse } from '../../utils/plugins'

const components = {
  FbxTextField
}

const VuePlugin = {
  install(Vue) {
    registerComponents(Vue, components)
  }
}

vueUse(VuePlugin)

export default VuePlugin
