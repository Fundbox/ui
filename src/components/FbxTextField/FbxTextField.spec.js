import { shallowMount } from '@vue/test-utils'
import FbxTextField from './FbxTextField.vue'
import VeeValidate from 'vee-validate'

const v = new VeeValidate.Validator()

describe('FbxTextField', () => {
  describe('snapshots', () => {
    it('renders the default', () => {
      const wrapper = shallowMount(FbxTextField, {
        propsData: {
          name: 'my-input',
          value: 'avocado',
          label: 'Favorite Food',
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders clear icon', () => {
      const wrapper = shallowMount(FbxTextField, {
        propsData: {
          name: 'my-input',
          value: 'avocado',
          clearIcon: true,
        },
        provide: () => ({
          $validator: v
        })
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('togglePassword', () => {
      it('changes the input type from password to text', () => {
        const wrapper = shallowMount(FbxTextField, {
          propsData: {
            name: 'my-input',
            value: 'avocado',
            type: 'text',
          },
        })

        wrapper.setData({
          type: 'password',
        })

        expect(wrapper.vm.type).toBe('password')

        wrapper.vm.togglePassword()

        expect(wrapper.vm.type).toBe('text')
      })

      it('changes the input type from text to password', () => {
        const wrapper = shallowMount(FbxTextField, {
          propsData: {
            name: 'my-input',
            value: 'avocado',
            type: 'text',
          },
        })
        expect(wrapper.vm.type).toBe('text')

        wrapper.vm.togglePassword()

        expect(wrapper.vm.type).toBe('password')
      })
    })

    describe('clearField', () => {
      it('emits an input event with an empty string', () => {
        const mockOnInput = jest.fn()
        const wrapper = shallowMount(FbxTextField, {
          propsData: {
            name: 'my-input',
            value: '42',
          },
          listeners: {
            input: mockOnInput,
          },
        })

        wrapper.vm.clearField()

        expect(mockOnInput).toHaveBeenCalled()
        expect(mockOnInput).toHaveBeenCalledWith('')
      })
    })

    describe('onInput', () => {
      it('emits the change event with the input value', () => {
        const mockOnInput = jest.fn()
        const wrapper = shallowMount(FbxTextField, {
          propsData: {
            name: 'my-input',
            value: '42',
          },
          listeners: {
            input: mockOnInput,
          },
        })

        wrapper.find('input').trigger('input')

        expect(mockOnInput).toHaveBeenCalled()
        expect(mockOnInput).toHaveBeenCalledWith('42')
      })
    })

    describe('onChange', () => {
      it('emits the change event with the input value', () => {
        const mockOnChange = jest.fn()
        const wrapper = shallowMount(FbxTextField, {
          propsData: {
            name: 'my-input',
            value: '42',
          },
          listeners: {
            change: mockOnChange,
          },
        })

        wrapper.find('input').trigger('change')

        expect(mockOnChange).toHaveBeenCalledTimes(1)
        expect(mockOnChange).toHaveBeenCalledWith('42')
      })
    })
  })
})
