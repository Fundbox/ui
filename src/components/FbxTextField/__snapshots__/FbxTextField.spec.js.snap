// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FbxTextField snapshots renders the default 1`] = `
<div class="fbx-text-field"><label class="fbx-text-field__label">Favorite Food</label>
  <div class="fbx-text-field__wrapper">
    <div class="fbx-text-field__input-wrapper"><input type="text" tabindex="0" class="fbx-text-field__input" data-mask="" data-previous-value="avocado" aria-required="false" aria-invalid="false">
      <!---->
    </div>
    <!---->
  </div>
</div>
`;
